compiler/dcalc/invariants.ml

```diff
@@ -0,0 +1,125 @@
+(* This file is part of the Catala compiler, a specification language for tax
+   and social benefits computation rules. Copyright (C) 2020 Inria, contributor:
+   Alain Delaët <alain.delaet--tixeuil@inria.fr>
+
+   Licensed under the Apache License, Version 2.0 (the "License"); you may not
+   use this file except in compliance with the License. You may obtain a copy of
+   the License at
+
+   http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+   WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+   License for the specific language governing permissions and limitations under
+   the License. *)
+
+open Shared_ast
+open Ast
+open Catala_utils
+
+type invariant_status = Fail | Pass | Ignore
+type invariant_expr = typed expr -> invariant_status
+
+let check_invariant (inv : string * invariant_expr) (p : typed program) : bool =
+  (* TODO: add a Program.fold_exprs to get rid of the reference 0:-)? *)
+  let result = ref true in
+  let name, inv = inv in
+  let _ = name in
+  let total = ref 0 in
+  let ok = ref 0 in
+  let p' =
+    Program.map_exprs p ~varf:Fun.id ~f:(fun e ->
+        (* let currente = e in *)
+        let rec f e =
+          let r =
+            match inv e with
+            | Ignore -> true
+            | Fail ->
+              Cli.error_format "%s failed in %s.\n\n %a" name
+                (Pos.to_string_short (Expr.pos e))
+                (Print.expr ~debug:true p.decl_ctx)
+                e;
+              incr total;
+              false
+            | Pass ->
+              incr ok;
+              incr total;
+              true
+          in
+          Expr.map_gather e ~acc:r ~join:( && ) ~f
+        in
+
+        let res, e' = f e in
+        result := res && !result;
+        e')
+  in
+  assert (Bindlib.free_vars p' = Bindlib.empty_ctxt);
+  Cli.result_print "Invariant %s\n   checked. result: [%d/%d]" name !ok !total;
+  !result
+
+(* Structural invariant: no default can have as type A -> B *)
+let invariant_default_no_arrow () : string * invariant_expr =
+  ( __FUNCTION__,
```

Comment:
It is a "macro" that gives the current function name. It is to avoid retyping the function name as exception name. https://v2.ocaml.org/api/Stdlib.html#VAL__FUNCTION__

Reply:
This "macro" could be used to in the Cli when describing flags.

Reply:
Neat, didn't know that.